### DISCUSSION TITLE
Is there a best practice for multiple elements?

### DISCUSSION THREAD

**Discussion by jrhager84** [thebuilder/react-intersection-observer#460] [Q&A]
Please forgive my ignorance, as I'm used to setting up an array of elements for the Intersection Observer API, and I'm not sure if that's the route to take here with motion and this package. Any resources or assistance would be greatly appreciated. 

I just have multiple elements that I want to add props to (e.g. 'up, down, fade, slide, etc') on different elements so I can use the same observer for all of them, just passing those animation types in each element. 

Thank you so much!

**Comment by thebuilder** [comment]
You should just use the library, creating a new hook for each element you need to observe. The library will reuse existing Intersection Observers as long as they have the same settings (threshold, etc). 

Not sure I follow how you intend to share props between different elements?

**Reply by jrhager84** [reply]
It would just get VERY messy creating a ref for *every* element, rather than an array of refs (passing it into a function for instance). Usually, I would query every selector, and push them into an array of targets with parameters.

How I did it before was I had a switch block in the Intersection Observer that would take props (duration, direction, threshold, etc) from the actual element using data- attributes. So I could pass an array of references with different parameters for each element, using only one Intersection Observer.

When an element would cross the threshold, it would fire the animation switch block, which would look at the associated parameters and trigger its corresponding animation etc.

Is that not possible with this library?

**Reply by thebuilder** [reply]
If you want to do this in the "React" way, this library is perfect for handling it. 
Consider making a `<AnimateIn>` component that takes your `duration`, `direction`, `threshold` or settings that's relevant, to the trigger the animation. Have it assign the `ref` to a `<div>` element it's controlling (and animating). 

See https://github.com/thebuilder/react-intersection-observer/blob/master/docs/Recipes.md#trigger-animations

```jsx
function AnimateIn(props) {
  const { ref, inView } = useInView({
    triggerOnce: true,
    threshold: props.threhold ?? 0,
    rootMargin: '-100px 0px',
  });

  return (
    <motion.div 
        ref={ref} 
        animate={{ opacity: inView ? 1 : 0 }} 
        transition={{ duration: props.duration ?? 1 }}
    >
      {props.children}
    </motion.div>
  );
}
```

**Reply by jrhager84** [reply]
Wouldn't I still have to set a ref individually for that wrapper? Seeing as how it takes the props on the instance (which wraps each element). I was thinking about using the InView HOC to wrap the entire app and then managing the ref array there, forwarding the ref down through the children. That way, I'd only have to have one instance (I know it's intelligent, but wrapping 15 components in this wrapper and then managing the refs can become unweilding)

If I'm not mistaken, it'll look in the refArray to see if any are inView, and then fire the appropriate action tied to the ref, correct? Maybe I'm just not understanding the 'react' way of doing this...

Something like this -> https://codepen.io/jrhager84/pen/vYNevpy?editors=1010

**Reply by thebuilder** [reply]
It sounds like you are trying to over complicate it. 

Your example would look something like this:

```jsx
<AnimateIn animation="slide-left">
    <h1>ONE</h1>
    <p>We are *not* staggering this.</p>
</AnimateIn >
<AnimateIn animation="slide-up" staggerChildren>
    <h1>TWO</h1>
    <p>We are going to stagger this.</p>
    <p>We are going to stagger this.</p>
    <p>We are going to stagger this.</p>
</AnimateIn>
<AnimateIn animation="slide-right">
    <h1>THREE</h1>
</AnimateIn >
```

The `<AnimateIn>` component would handle its own `useInView`, and trigger the animation as needed.
For your usecase, combining it with https://www.framer.com/motion/ would make a lot of sense, since it can handle triggering nested animations and variants (like your `slide-left`, `slide-right`).

**Reply by jrhager84** [reply]
So does the useInView create and manage its own refs?

**Reply by thebuilder** [reply]
Yes, you just have to assign it to a an element, and it will do the rest.

**Reply by jrhager84** [reply]
Then if *definitely* seems as though I missed the entire point of this. That's the issue I was facing - Creating and managing the refs of every element outside of the component seemed ridiculous to me. The fact that I can use the ref of inView to automagically create a ref is exactly what I need. Thanks! I'll dive in tonight! :)

**Reply by michael-mcgreal** [reply]
Very helpful example – thanks as well!
